File: util/src/graph.rs
use std::rc::Rc;

mod adjacency_list;
mod dijkstra;

pub use adjacency_list::*;
pub use dijkstra::*;

pub trait Graph {
    type Element;

    fn edges(&self, source: &Self::Element) -> &Vec<Edge<Self::Element>>;

    fn weight(&self, source: &Self::Element, destination: &Self::Element) -> Option<f32>;

    fn insert(&mut self, data: Self::Element);

    fn add_directed_edge(
        &mut self,
        source: &Rc<Self::Element>,
        destination: &Rc<Self::Element>,
        weight: Option<f32>,
    );

    fn add_undirected_edge(
        &mut self,
        source: &Rc<Self::Element>,
        destination: &Rc<Self::Element>,
        weight: Option<f32>,
    ) {
        self.add_directed_edge(source, destination, weight);
        self.add_directed_edge(destination, source, weight);
    }

    fn add_edge(
        &mut self,
        edge: EdgeType,
        source: &Rc<Self::Element>,
        destination: &Rc<Self::Element>,
        weight: Option<f32>,
    ) {
        match edge {
            EdgeType::Directed => self.add_directed_edge(source, destination, weight),
            EdgeType::Undirected => self.add_undirected_edge(source, destination, weight),
        }
    }
}

#[derive(Debug)]
pub struct Edge<Element> {
    pub source: Rc<Element>,
    pub destination: Rc<Element>,
    pub weight: Option<f32>,
}

pub enum EdgeType {
    Directed,
    Undirected,
}

File: Sources/Base/Graph.swift
public protocol Graph {
    associatedtype Element

    func edges(from source: Element) -> [Edge<Element>]
    func weight(from source: Element, to destination: Element) -> Double?

    mutating func insert(_ element: Element, edges: [Edge<Element>])
    mutating func addDirectedEdge(from source: Element, to destination: Element, weight: Double?)
    mutating func addUndirectedEdge(from source: Element, to destination: Element, weight: Double?)
}

extension Graph {
    public mutating func addUndirectedEdge(from source: Element, to destination: Element, weight: Double?) {
        addDirectedEdge(from: source, to: destination, weight: weight)
        addDirectedEdge(from: destination, to: source, weight: weight)
    }
}

public struct Edge<Element> {
    public var source: Element
    public var destination: Element
    public var weight: Double?
}

public struct AdjacencyList<T: Hashable> {
    private var inner: [T: [Edge<T>]] = [:]

    public var nodes: [T] {
        Array(inner.keys)
    }

    public var count: Int {
        inner.sum(of: \.value.count)
    }

    public var isEmpty: Bool {
        inner.isEmpty
    }

    public init() {}
}

extension AdjacencyList: Graph {
    public func edges(from source: T) -> [Edge<T>] {
        inner[source] ?? []
    }

    public func weight(from source: T, to destination: T) -> Double? {
        edges(from: source)
            .first { $0.destination == destination }
            .flatMap { $0.weight }
    }

    public mutating func insert(_ element: T, edges: [Edge<T>] = []) {
        if contains(element) {
            return
        }

        inner[element] = edges
    }

    public mutating func addDirectedEdge(from source: T, to destination: T, weight: Double?) {
        insert(source)

        let edge = Edge(source: source,
                        destination: destination,
                        weight: weight)

        inner[source]!.append(edge)
    }
}

extension AdjacencyList: Sequence {
    public func makeIterator() -> Dictionary<T, Array<Edge<T>>>.Keys.Iterator {
        inner.keys.makeIterator()
    }
}

extension AdjacencyList: ExpressibleByArrayLiteral {
    public init(arrayLiteral elements: (T, [Edge<T>])...) {
        self.init()
        for element in elements {
            insert(element.0, edges: element.1)
        }
    }
}
